device/iot-device-samples/custom-sas-token-provider-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/CustomSasTokenProviderSample.java

```diff
@@ -0,0 +1,370 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+package samples.com.microsoft.azure.sdk.iot;
+
+import com.microsoft.azure.sdk.iot.device.*;
+import com.microsoft.azure.sdk.iot.device.transport.IotHubConnectionStatus;
+import org.apache.commons.codec.binary.Base64;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.List;
+
+
+/**
+ * This sample demonstrates how to configure your device client to use a custom SAS token provider instead of
+ * directly providing it the device's symmetric key.
+ */
+public class CustomSasTokenProviderSample
+{
+    private static final int D2C_MESSAGE_TIMEOUT = 2000; // 2 seconds
+    private static List<String> failedMessageListOnClose = new ArrayList(); // List of messages that failed on close
+
+    /**
+     * Helper class for turning symmetric keys into SAS tokens. It also provides some helpful functions around
+     * if this token should be renewed.
+     */
+    protected static class SasToken
+    {
+        private static final String RAW_SIGNATURE_FORMAT = "%s\n%s";
```

Comment:
I wonder if these should be constants in the SDK so someone can just refer to them instead of having to know or refer to your sample.

Reply:
It's tempting, but we don't have a great place to put it. None of our "public" classes are directly tied to our connection string/sas token logic

Reply:
It could be a standalone constants class